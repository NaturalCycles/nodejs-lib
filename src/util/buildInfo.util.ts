import {
  _filterUndefinedValues,
  AnyObject,
  BuildInfo,
  localTimeOrNow,
  UnixTimestampNumber,
} from '@naturalcycles/js-lib'
import { fs2 } from '../fs/fs2'
import {
  gitCurrentBranchName,
  gitCurrentCommitSha,
  gitCurrentCommitTimestamp,
  gitCurrentRepoName,
} from './git.util'

export interface GenerateBuildInfoOptions {
  /**
   * If set - this timestamp will be used, instead of "current time".
   */
  overrideTimestamp?: UnixTimestampNumber
}

export function generateBuildInfo(opt: GenerateBuildInfoOptions = {}): BuildInfo {
  const now = localTimeOrNow(opt.overrideTimestamp)
  const ts = now.unix()

  const rev = gitCurrentCommitSha()
  const branchName = gitCurrentBranchName()
  const repoName = gitCurrentRepoName()
  const tsCommit = gitCurrentCommitTimestamp()

  const ver = [now.toStringCompact(), repoName, branchName, rev].join('_')

  let { APP_ENV: env } = process.env

  if (!env) {
    // Attempt to read `envByBranch` from package.json root
    try {
      if (fs2.pathExists('package.json')) {
        const packageJson = fs2.readJson<AnyObject>('package.json')
        env = packageJson?.['envByBranch']?.[branchName] || packageJson?.['envByBranch']?.['*']
      }
    } catch {}
  }

  return _filterUndefinedValues({
    ts,
    tsCommit,
    repoName,
    branchName,
    rev,
    ver,
    env,
  })
}
