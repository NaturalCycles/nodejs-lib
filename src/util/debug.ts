import debug from 'debug'
import { IDebugger } from 'debug'
import * as path from 'path'

export function getDebug (filename: string): IDebugger {
  return debug(getDebugNamespaceFromFilename(filename))
}

function getDebugNamespaceFromFilename (filename: string): string {
  // drop last part after dot
  const a = path.basename(filename).split('.')
  a.pop()
  return a.join('.')
}
