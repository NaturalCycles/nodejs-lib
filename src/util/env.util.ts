import type { ValuesOf } from '@naturalcycles/js-lib'
import 'dotenv/config' // ensure .env is read before requiring keys
import * as fs from 'fs-extra'

/**
 * @example
 *
 * const {a, b} = requreEnvKeys(['a', 'b'])
 *
 * Will throw if any of the passed keys is not defined.
 */
export function requireEnvKeys<T extends readonly string[]>(
  ...keys: T
): { [k in ValuesOf<T>]: string } {
  // eslint-disable-next-line unicorn/no-array-reduce
  return keys.reduce((r, k) => {
    const v = process.env[k]
    if (!v) throw new Error(`${k} env variable is required, but missing`)
    r[k] = v
    return r
  }, {} as { [k in ValuesOf<T>]: string })
}

export function requireFileToExist(filePath: string): void {
  if (!fs.pathExistsSync(filePath)) {
    throw new Error(`Required file should exist: ${filePath}`)
  }
}
