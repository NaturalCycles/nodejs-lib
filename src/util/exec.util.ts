import type { ProcessEnvOptions, SpawnOptions } from 'node:child_process'
import * as cp from 'node:child_process'

export async function execVoidCommand(
  cmd: string,
  args: string[] = [],
  opt: SpawnOptions = {},
): Promise<void> {
  logExec(cmd, args, opt)

  await new Promise<void>(resolve => {
    const p = cp.spawn(cmd, [...args], {
      stdio: 'inherit',
      // shell: true,
      ...opt,
      env: {
        ...process.env,
        ...opt.env,
      },
    })

    p.on('close', code => {
      if (!code) return resolve()
      console.log(`${cmd} exited with code ${code}`)
      process.exit(code)
    })
  })
}

export function execVoidCommandSync(
  cmd: string,
  args: string[] = [],
  opt: SpawnOptions = {},
): void {
  logExec(cmd, args, opt)

  const r = cp.spawnSync(cmd, [...args], {
    encoding: 'utf8',
    stdio: 'inherit',
    // shell: true, // removing shell breaks executing `tsc`
    ...opt,
    env: {
      ...process.env,
      ...opt.env,
    },
  })

  if (r.status) {
    console.log(`${cmd} exited with code ${r.status}`)
    process.exit(r.status)
  }

  if (r.error) {
    console.log(r.error)
    process.exit((r.error as NodeJS.ErrnoException).errno || 1)
  }
}

function logExec(cmd: string, args: string[] = [], opt: ProcessEnvOptions = {}): void {
  const cmdline = [
    ...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('=')),
    cmd,
    ...args,
  ].join(' ')

  console.log(cmdline)
}
