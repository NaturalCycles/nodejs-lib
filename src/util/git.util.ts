import * as cp from 'node:child_process'
import * as path from 'node:path'
import type { UnixTimestampNumber } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors'

export function getLastGitCommitMsg(): string {
  return execSync('git log -1 --pretty=%B')
}

export function commitMessageToTitleMessage(msg: string): string {
  const firstLine = msg.split('\n')[0]!
  const [preTitle, title] = firstLine.split(': ')
  return title || preTitle!
}

export function gitHasUncommittedChanges(): boolean {
  // git diff-index --quiet HEAD -- || echo "untracked"
  try {
    cp.execSync('git diff-index --quiet HEAD --', {
      encoding: 'utf8',
    })
    return false
  } catch {
    return true
  }
}

/**
 * @returns true if there were changes
 */
export function gitCommitAll(msg: string): boolean {
  // git commit -a -m "style(lint-all): $GIT_MSG" || true
  const cmd = `git commit -a --no-verify -m "${msg}"`
  // const cmd = `git`
  // const args = ['commit', '-a', '--no-verify', '-m', msg]

  console.log(grey(cmd))

  try {
    cp.execSync(cmd, {
      stdio: 'inherit',
    })
    return true
  } catch {
    return false
  }
}

/**
 * @returns true if there are not pushed commits.
 */
export function gitIsAhead(): boolean {
  // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
  const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`

  const stdout = execSync(cmd)

  // console.log(`gitIsAhead: ${stdout}`)
  return Number(stdout) > 0
}

export function gitPull(): void {
  const cmd = 'git pull'
  try {
    cp.execSync(cmd, {
      stdio: 'inherit',
    })
  } catch {}
}

export function gitPush(): void {
  // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
  let cmd = 'git push'

  const branchName = gitCurrentBranchName()

  if (branchName) {
    cmd += ` --set-upstream origin ${branchName}`
  }

  console.log(grey(cmd))

  cp.execSync(cmd, {
    stdio: 'inherit',
  })
}

export function gitCurrentCommitSha(full = false): string {
  const sha = execSync('git rev-parse HEAD')
  return full ? sha : sha.slice(0, 7)
}

export function gitCurrentCommitTimestamp(): UnixTimestampNumber {
  return Number(execSync('git log -1 --format=%ct'))
}

export function gitCurrentBranchName(): string {
  return execSync('git rev-parse --abbrev-ref HEAD')
}

export function gitCurrentRepoName(): string {
  const originUrl = execSync('git config --get remote.origin.url')
  return path.basename(originUrl, '.git')
}

function execSync(cmd: string): string {
  return cp
    .execSync(cmd, {
      encoding: 'utf8',
      // stdio: 'inherit', // no, otherwise we don't get the output returned
    })
    .trim()
}
