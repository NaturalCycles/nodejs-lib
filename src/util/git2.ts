import cp from 'node:child_process'
import path from 'node:path'
import type { UnixTimestampNumber } from '@naturalcycles/js-lib'
import { grey } from '../colors/colors'
import { exec2 } from './exec2'

/**
 * Set of utility functions to work with git.
 */
class Git2 {
  getLastGitCommitMsg(): string {
    return exec2.exec('git log -1 --pretty=%B')
  }

  commitMessageToTitleMessage(msg: string): string {
    const firstLine = msg.split('\n')[0]!
    const [preTitle, title] = firstLine.split(': ')
    return title || preTitle!
  }

  gitHasUncommittedChanges(): boolean {
    // git diff-index --quiet HEAD -- || echo "untracked"
    try {
      cp.execSync('git diff-index --quiet HEAD --', {
        encoding: 'utf8',
      })
      return false
    } catch {
      return true
    }
  }

  /**
   * Returns true if there were changes
   */
  gitCommitAll(msg: string): boolean {
    // git commit -a -m "style(lint-all): $GIT_MSG" || true
    const cmd = `git commit -a --no-verify -m "${msg}"`
    // const cmd = `git`
    // const args = ['commit', '-a', '--no-verify', '-m', msg]
    console.log(grey(cmd))

    try {
      cp.execSync(cmd, {
        stdio: 'inherit',
      })
      return true
    } catch {
      return false
    }
  }

  /**
   * @returns true if there are not pushed commits.
   */
  gitIsAhead(): boolean {
    // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
    const stdout = exec2.exec(cmd)
    // console.log(`gitIsAhead: ${stdout}`)
    return Number(stdout) > 0
  }

  gitPull(): void {
    const cmd = 'git pull'
    try {
      cp.execSync(cmd, {
        stdio: 'inherit',
      })
    } catch {}
  }

  gitPush(): void {
    // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
    let cmd = 'git push'

    const branchName = this.gitCurrentBranchName()

    if (branchName) {
      cmd += ` --set-upstream origin ${branchName}`
    }

    exec2.spawn(cmd, { logStart: true })
  }

  gitCurrentCommitSha(full = false): string {
    const sha = exec2.exec('git rev-parse HEAD')
    return full ? sha : sha.slice(0, 7)
  }

  gitCurrentCommitTimestamp(): UnixTimestampNumber {
    return Number(exec2.exec('git log -1 --format=%ct'))
  }

  gitCurrentBranchName(): string {
    return exec2.exec('git rev-parse --abbrev-ref HEAD')
  }

  gitCurrentRepoName(): string {
    const originUrl = exec2.exec('git config --get remote.origin.url')
    return path.basename(originUrl, '.git')
  }
}

export const git2 = new Git2()
