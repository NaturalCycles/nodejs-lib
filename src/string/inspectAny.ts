import { inspect, InspectOptions } from 'util'
import { StringifyAnyOptions, _stringifyAny, JsonStringifyFunction } from '@naturalcycles/js-lib'

export interface InspectAnyOptions extends StringifyAnyOptions, InspectOptions {}

const INSPECT_OPT: InspectOptions = {
  breakLength: 80, // default: ??
  depth: 10, // default: 2
}

/**
 * Just a convenience export of a const that fulfills the JsonStringifyFunction interface.
 */
export const inspectAnyStringifyFn: JsonStringifyFunction = obj => inspectAny(obj)

/**
 * Transforms ANY to human-readable string (via util.inspect mainly).
 * Safe (no error throwing).
 *
 * Correclty prints Errors, AppErrors, ErrorObjects: error.message + \n + inspect(error.data)
 *
 * Enforces max length (default to 10_000, pass 0 to skip it).
 *
 * Logs numbers as-is, e.g: `6`.
 * Logs strings as-is (without single quotes around, unlike default util.inspect behavior).
 * Otherwise - just uses util.inspect() with reasonable defaults.
 *
 * Returns 'empty_string' if empty string is passed.
 * Returns 'undefined' if undefined is passed (default util.inspect behavior).
 *
 * Based on `_stringifyAny` from `js-lib`, just replaced `JSON.stringify` with `util.inspect`.
 */
export function inspectAny(obj: any, opt: InspectAnyOptions = {}): string {
  // Inspect handles functions better
  if (typeof obj === 'function') {
    return inspect(obj, {
      ...INSPECT_OPT,
      ...opt,
    })
  }

  return _stringifyAny(obj, {
    ...opt,
    stringifyFn: obj =>
      inspect(obj, {
        ...INSPECT_OPT,
        ...opt,
      }),
  })
}
