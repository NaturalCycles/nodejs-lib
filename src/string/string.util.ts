import { AppError, _isErrorObject } from '@naturalcycles/js-lib'
import { inspect } from 'util'

export interface InspectIfPossibleOptions extends NodeJS.InspectOptions {
  /**
   * @default 10_000
   */
  maxLen?: number

  /**
   * @default false
   * Set to true to not print Error.stack (keeping just Error.message).
   */
  noErrorStack?: boolean
}

/**
 * Transforms ANY to human-readable string (via util.inspect mainly).
 * Safe (no error throwing).
 *
 * Correclty prints Errors, AppErrors, ErrorObjects: error.message + \n + inspect(error.data)
 *
 * Enforces max length (default to 10_000, pass 0 to skip it).
 *
 * Logs numbers as-is, e.g: `6`.
 * Logs strings as-is (without single quotes around, unlike default util.inspect behavior).
 * Otherwise - just uses util.inspect() with reasonable defaults.
 *
 * Returns 'empty_string' if empty string is passed.
 * Returns 'undefined' if undefined is passed (default util.inspect behavior).
 */
export function inspectAny(obj: any, opt: InspectIfPossibleOptions = {}): string {
  let s: string

  if (obj instanceof Error) {
    // Stack includes message
    s = (!opt.noErrorStack && obj.stack) || [obj?.name, obj.message].filter(Boolean).join(': ')

    if (obj instanceof AppError) {
      s = [s, Object.keys(obj.data).length > 0 && inspectAny(obj.data, opt)]
        .filter(Boolean)
        .join('\n')
    }
  } else if (_isErrorObject(obj)) {
    s = [obj.message, Object.keys(obj.data).length > 0 && inspectAny(obj.data, opt)]
      .filter(Boolean)
      .join('\n')
  } else if (typeof obj === 'string') {
    s = obj.trim() || 'empty_string'
  } else {
    s = inspect(obj, {
      breakLength: 80, // default: ??
      depth: 6, // default: 2
      ...opt,
    })
  }

  // Handle maxLen
  if (opt.maxLen && s.length > opt.maxLen) {
    s = s.substr(0, opt.maxLen) + `... ${Math.ceil(s.length / 1024)} KB message truncated`
  }

  return s
}
