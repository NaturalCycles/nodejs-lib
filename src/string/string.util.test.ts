import { expectResults, mockAllKindsOfThings } from '@naturalcycles/dev-lib/dist/testing'
import { _range } from '@naturalcycles/js-lib'
import { inspectAny } from '../index'
import { jsonParseIfPossible } from './string.util'

test('inspectAny', () => {
  expectResults(
    v => inspectAny(v, { noErrorStack: true }),
    mockAllKindsOfThings(),
  ).toMatchSnapshot()
})

test('inspectAny maxLen', () => {
  const obj = _range(1, 1000).join(',')
  expect(inspectAny(obj, { maxLen: 10 })).toBe(`1,2,3,4,5,... 4 KB message truncated`)
})

test('jsonParseIfPossible', () => {
  expectResults(v => jsonParseIfPossible(v), mockAllKindsOfThings()).toMatchSnapshot()
})
