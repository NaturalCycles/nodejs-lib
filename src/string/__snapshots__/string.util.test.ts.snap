// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspectAny 1`] = `
Map {
  undefined => "undefined",
  null => "null",
  true => "true",
  false => "false",
  NaN => "NaN",
  Infinity => "Infinity",
  "" => "empty_string",
  " " => "empty_string",
  "ho ho ho" => "ho ho ho",
  "{" => "{",
  "{\\"a\\": \\"a1\\"}" => "{\\"a\\": \\"a1\\"}",
  "[5, null, \\"s\\"]" => "[5, null, \\"s\\"]",
  Array [
    5,
    null,
    "s",
  ] => "[ 5, null, 's' ]",
  15 => "15",
  Array [
    1,
    2,
    3,
  ] => "[ 1, 2, 3 ]",
  Object {
    "a": "a1",
    "b": "b1",
    "c": Object {
      "d": 25,
      "e": undefined,
    },
  } => "{ a: 'a1', b: 'b1', c: { d: 25, e: undefined } }",
  Object {
    "documentation_url": "https://developer.github.com/v3/git/refs/#create-a-reference",
    "message": "Reference already exists",
  } => "{
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
}",
  [Function] => "[Function: abc]",
  /i am regex, who are you\\?/ => "/i am regex, who are you?/",
  Map {
    "a" => "b",
  } => "Map { 'a' => 'b' }",
  Set {
    "a",
    "b",
    "c",
  } => "Set { 'a', 'b', 'c' }",
  [Error: some err msg] => "Error: some err msg",
  [AppError: app error] => "AppError: app error
{ k1: 'v1' }",
  Symbol(a_am_symbol) => "Symbol(a_am_symbol)",
  1984-06-21T00:00:00.000Z => "1984-06-21T00:00:00.000Z",
}
`;
